### PR TITLE
Analytics Datapoint Refactoring

### PR DESCRIPTION
**Pull request by aaemnnosttv**
## Summary

Addresses issue #450

## Relevant technical choices
* Refactored `withData` to provide the request object to the child component as a new `requestData` prop to allow for request-specific handling  
_You can see why this is necessary in usage of `toState` in `dashboard/dashboard-widget-top-level.js`_

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by aaemnnosttv** [comment]
@felixarntz  - one thing that is a bit odd here is how `dimensions` are passed as a string (comma-separated) and everything else is using normal object structures. Should we update this to expect the parameter to be sent as an array instead?

**Comment by aaemnnosttv** [comment]
image

**Comment by felixarntz** [comment]
@aaemnnosttv 
> one thing that is a bit odd here is how dimensions are passed as a string (comma-separated) and everything else is using normal object structures. Should we update this to expect the parameter to be sent as an array instead?

Yes, that totally makes sense. I think I didn't phrase this correctly, we should probably expect an array here as baseline, but also support a string of values (that's how core handles simple scalar lists in their endpoints too).

File: assets/js/components/higherorder/withdata.js [hunk 1]
```diff
@@ -145,14 +145,15 @@ const withData = (
...
+			 * @param {Object} requestData  The data object for the request.
 			 */
-			const handleReturnedData = ( returnedData, datapoint, identifier ) => {
+			const handleReturnedData = ( returnedData, datapoint, requestData ) => {
 				// If available, `handleDataError` will be called for errors (with a string) and empty data.
 				const {
 					handleDataError,
 					handleDataSuccess,
 				} = this.props;
+				const { identifier } = requestData;
```
Comment (felixarntz) [inline assets/js/components/higherorder/withdata.js]: While we're at it, should we also remove the `datapoint` parameter and here get it from `requestData`?

Comment (aaemnnosttv) [reply assets/js/components/higherorder/withdata.js]: Done!

**Comment by aaemnnosttv** [comment]
> it's only used to get the new toState function, so if we need it, we should only pass that.

Sounds good.

> the new toState is only to cater for one very specific scenario which seems not optimal.

It's not specific to Analytics; the same problem is coming up in the Adsense module as well where state is set differently using the response based on the datapoint:

https://github.com/google/site-kit-wp/blob/1d62eefa792e6e1e5448a66e1bfad95a3e4a2367/assets/js/modules/adsense/dashboard/dashboard-widget-estimate-earnings.js#L84-L104

**Comment by aaemnnosttv** [comment]
@felixarntz Due to the shared dependency regarding the change to `withData` #453 will be blocked until this is merged.

**Pull request merged**